File: src/1273-Delete-Tree-Nodes/1273.py
class Solution:
    def deleteTreeNodes(self, n: int, parent: List[int], value: List[int]) -> int:
        for i in range(n-1, 0, -1):
            value[parent[i]] += value[i]
        
        zeros, isZero = 0, [0] * n
        for i in range(n):
            if parent[i] > 0 and isZero[parent[i]] or value[i] == 0:
                isZero[i] = 1
                zeros += 1
        return n - zeros

File: src/1273-Delete-Tree-Nodes/1273.cpp
class Solution 
{
public:
    int deleteTreeNodes(int n, vector<int>& parent, vector<int>& value) 
    {
        for (int i = n - 1; i > 0; i--) value[parent[i]] += value[i];
        int zeros = 0;
        bool isZero[n] = {};
        for (int i = 0; i < n; i++)
        {
            if (parent[i] > 0 && isZero[parent[i]] || value[i] == 0)
            {
                isZero[i] = true; zeros++;
            }
        }
        return n - zeros;
    }
};